backend/services/implementations/statisticsService.ts

```diff
@@ -0,0 +1,147 @@
+import { getErrorMessage } from "../../utilities/errorUtils";
+import logger from "../../utilities/logger";
+import { ISchoolService, SchoolResponseDTO } from "../interfaces/schoolService";
+import {
+  IStatisticsService,
+  AggregateTestSessionDTO,
+  StatisticsResponseDTO,
+} from "../interfaces/statisticsService";
+import { ITestService, TestResponseDTO } from "../interfaces/testService";
+import { ITestSessionService } from "../interfaces/testSessionService";
+import IUserService from "../interfaces/userService";
+
+const Logger = logger(__filename);
+
+class StatisticsService implements IStatisticsService {
+  /* eslint-disable class-methods-use-this */
+  testSessionService: ITestSessionService;
+
+  testService: ITestService;
+
+  userService: IUserService;
+
+  schoolService: ISchoolService;
+
+  constructor(
+    testSessionService: ITestSessionService,
+    testService: ITestService,
+    userService: IUserService,
+    schoolService: ISchoolService,
+  ) {
+    this.testSessionService = testSessionService;
+    this.testService = testService;
+    this.userService = userService;
+    this.schoolService = schoolService;
+  }
+
+  async getAverageScoresBySchool(
+    testId: string,
+  ): Promise<Array<StatisticsResponseDTO>> {
+    let statisticsDTOs: Array<StatisticsResponseDTO> = [];
+
+    try {
+      const statistics: AggregateTestSessionDTO[] = await MgTestSession.aggregate(
+        [
+          // Stage 1: Filter Test Session documents by test id
+          {
+            $match: { test: Types.ObjectId(testId) },
+          },
+          // Stage 2: Group remaining documents by school id and add aggregate fields
+          {
+            $unwind: "$results",
+          },
+          {
+            $group: {
+              _id: "$school",
+              averageScore: { $avg: "$results.score" },
+              resultsPerTestArr: {
+                $push: {
+                  $map: {
+                    input: "$results.breakdown",
+                    as: "result",
+                    in: { $cond: ["$$result", 1, 0] },
+                  },
+                },
+              },
+            },
+          },
+          // Stage 3: Sort documents by averageScore in descending order
+          {
+            $sort: { averageScore: -1 },
+          },
+        ],
+      );
+
+      /* eslint-disable no-param-reassign */
+      statistics.forEach((statistic: AggregateTestSessionDTO) => {
+        statistic.averageScoresByQuestions = this.computeAverageScorePerQuestion(
+          statistic.resultsPerTestArr,
+        );
+      });
+
+      statisticsDTOs = await this.mapStatisticsToStatisticsDTOs(
+        testId,
+        statistics,
+      );
+    } catch (error: unknown) {
+      Logger.error(
+        `Failed to get average scores by school. Reason = ${getErrorMessage(
+          error,
+        )}`,
+      );
+      throw error;
+    }
+
+    return statisticsDTOs;
+  }
+
+  private computeAverageScorePerQuestion(
```

Comment:
would prefer doing this directly in the aggregation pipeline but haven't found a good way to implement :(.  the closest solution would be using `$zip` but there's this [bug](https://jira.mongodb.org/browse/SERVER-27964?page=com.atlassian.streams.streams-jira-plugin%3Aactivity-stream-issue-tab) that restricts the function's use to arrays with fixed dimensions.

Reply:
essentially we want to transpose an array of test results (i.e. `[result1, result2, result3,...]` with each result being an array of student answers) to `[question1_answers, question2_answers, ...]` with each entry being all answers for each question

Reply:
I think it's okay to leave out of aggregation pipeline, as long as the meaning is clear! Which I think it is here.